### DISCUSSION TITLE
Support for WebLN on the payment page

### DISCUSSION THREAD

**Discussion by bumi** [btcpayserver/btcpayserver#2426] [Ideas / Feature Requests]
"WebLN is a library and set of specifications for lightning apps and client providers to facilitate communication between apps and users' lightning nodes in a secure way. It provides a programmatic, permissioned interface for letting applications ask users to send payments, generate invoices to receive payments, and much more."

It seems the payment page currently does not implement the WebLN spec but simply provides a `lightning:` link. 
The WebLN spec provides a simple specification on how to request payments from a user programmatically. 

It can be added unobtrusively and enhances the checkout experience for people with a Browser supporting WebLN.

**Comment by ghost** [comment]
idk, [it says](https://github.com/joule-labs/webln) `This spec is in early stages, and is subject to change.`, but the last commit to the repo was June 2019...

I don't know where else it is implemented, other than [Joule](https://github.com/joule-labs/joule-extension/) [which is also not really worked on anymore...]

**Reply by bumi** [reply]
e.g. breez and [bluewallet](https://github.com/BlueWallet/BlueWallet/blob/0ec80f0616db9efc5204e6758630171fbe6985b7/screen/lnd/browser.js#L108) support it. (e are also working on a general [lighting browser extension](github.com/bumi/lightning-browser-extension/) with the goal allowing programmatic payments. (shameless plug)
With the recent implementations around payments for podcast I think it is important that we work towards a common interface to interact with the lightning network from the browser. 
UX improvements can be quite big when it comes to online content. (why stare at a QR code to read a newspaper article? - and "streaming" money to BTCPay then can become possible from the browser) 

WebLN defines a very small interface for common LN interactions and even if the repo has not been updated it works perfectly fine as it for example simply describes a JS interface on how to pass an invoice to request a payment. (

For BTCPay an initial implementation is very minimal and is basically as simple as a event listener on the current invoice links. Automatically calling WebLN (if enabled) and showing the QR code only as a fallback could also be done later. 

What are your concerns regarding adding support for the interface to BTCPAY?

edit: to look at the merchant side. e.g. yalls or bitrefill support WebLN.

**Reply by ghost** [reply]
I'm happy to hear that there's such an uptick in usage and implementations! Well done.

I agree that such a UX flow is useful. I can't speak for the technical feasibility / concerns of adding it to btcpay.

Potentially, this could be done in a plugin, but maybe it's a "core feature" worth to put in the main software.

**Reply by bumi** [reply]
If somebody could point me to the code of the checkout screens I could try to make a PR showing how it could work.

**Reply by dennisreimann** [reply]
@bumi The template and JS is located [here](https://github.com/btcpayserver/btcpayserver/blob/master/BTCPayServer/Views/Invoice/Checkout.cshtml), it's basically a Vue.js app. In general we try to be rather conservative with changes to the checkout. Let me know if you have questions or need help implementing a proof of concept :)

**Reply by bumi** [reply]
thanks for the pointer. where does it show the lightning invoice?

Is there an API that I maybe do not even have to go through the checkout page? and the checkout page is only the fallback if webln is not available?

**Reply by pavlenex** [reply]
@bumi There's an API https://docs.btcpayserver.org/API/Greenfield/v1/

**Comment by kingonly** [comment]
What's the status of this?
Merchants using btcpay are asking us to add them to the Breez marketplace, but we need them to be webln compatible.

**Reply by dennisreimann** [reply]
Can you elaborate on what exactly is needed? I remember working on integrating WebLN, but to us there was no direct benefit, because the general use cases seemed to work well enough with the `lightning:` scheme and the existing handlers.

**Reply by kingonly** [reply]
It's super simple: integrate [this](https://www.webln.guide/building-lightning-apps/webln-reference/webln.makeinvoice) in your checkout screen. 

It has nothing to do with `lightning:` scheme. This is for allowing webln hosts (Breez, Alby, Blixt) to seamlessly pay BTC Pay merchants. Bitrefill, LNM, Wavlake, FixedFloat, Boltz, TBC, Geyser, sms4sats, Lightsats, LNCal are just a few examples of sites that have makeInvoice integration and accept funds using webln.

This will allow us (Breez) to embed BTC Pay widgets in our marketplace.
